### ISSUE TITLE
Weighted average is not used in FedAvg

### ISSUE THREAD

**Issue by bird-two** [SMILELab-FL/FedLab#305]
Thanks for your valuable time for cheking this issue.

**Describe the bug**
Weighted average is not used in FedAvg (see [fedavg.py](https://github.com/SMILELab-FL/FedLab/blob/master/fedlab/contrib/algorithm/fedavg.py))

**Details**
My understanding on this package. 
1. `global_update()` is performed in the server.

Then we show the details.
First, we found that the server inherite the `SyncServerHandler` without more modifications (see line 13 [fedavg.py](https://github.com/SMILELab-FL/FedLab/blob/master/fedlab/contrib/algorithm/fedavg.py)). 

Then we see that FedAvg use a unweighted average on the model parameters at lines 106-109 in [basic_server.py](https://github.com/SMILELab-FL/FedLab/blob/master/fedlab/contrib/algorithm/basic_server.py). 

We note that class `FedAvgClientTrainer()` at lines 25-32 in [fedavg.py](https://github.com/SMILELab-FL/FedLab/blob/master/fedlab/contrib/algorithm/fedavg.py) is rewriten with a new function `global_update()`, yet this contradicts with my understanding 1. So maybe the function `global_update()` should be added in the class `FedAvgServerHandler()`.

In addition, we note that the _global learning rate_ (see ref [1,2]) has been widely adopted in FedAvg, so it would be better to add the global learning rate in FedAvg.

[1] Karimireddy, S. P., Kale, S., Mohri, M., Reddi, S., Stich, S., & Suresh, A. T. (2020, November). Scaffold: Stochastic controlled averaging for federated learning. In International Conference on Machine Learning (pp. 5132-5143). PMLR.

[2] Wang, J., Charles, Z., Xu, Z., Joshi, G., McMahan, H. B., Al-Shedivat, M., ... & Zhu, W. (2021). A field guide to federated optimization. arXiv preprint arXiv:2107.06917.

**Comment by Zengdun-cs** [comment]
Thank you very much for your careful code review.

1. weight aggregation can be launched by passing the weights of clients. Please see the [code](https://github.com/SMILELab-FL/FedLab/blob/master/fedlab/utils/aggregator.py).
2. Global iteration with learning rate has been implemented in our code base, e.g., [SCAFFOLD](https://github.com/SMILELab-FL/FedLab/blob/master/fedlab/contrib/algorithm/scaffold.py).

Thanks again for your attention and patience. We appreciate your efforts to improve our repo. Besides, we are also open to code contributions.

**Comment by bird-two** [comment]
Thanks for your response.

For clarity, we use _weight_ to denote the weight of each client's mode parameter when making aggregation (or average) and _model parameters_ to denote the model parameters.

1. Yes, we can use weighted aggregation on the model parameters by using the [code](https://github.com/SMILELab-FL/FedLab/blob/master/fedlab/utils/aggregator.py). Yet, we note that for [fedavg.py](https://github.com/SMILELab-FL/FedLab/blob/master/fedlab/contrib/algorithm/fedavg.py), **the repo does not use weighted average** (vanilla average with each client's weight equal to each other is used here), since the repo does not pass the clients' weights to the function `fedavg_aggregate` (see lines 106-109 in [basic_server](https://github.com/SMILELab-FL/FedLab/blob/master/fedlab/contrib/algorithm/basic_server.py)).
2. For the 2-nd point, I get it well. Yet, I still think it need to be added into FedAvg.

**Comment by bird-two** [comment]
To validate this issue. I modify lines 13-15 in [fedavg.py](https://github.com/SMILELab-FL/FedLab/blob/master/fedlab/contrib/algorithm/fedavg.py) with the code below.
```
class FedAvgServerHandler(SyncServerHandler):
    def global_update(self, buffer):
        parameters_list = [ele[0] for ele in buffer]
        weights = [ele[1] for ele in buffer]
        serialized_parameters = Aggregators.fedavg_aggregate(
            parameters_list, weights)
        SerializationTool.deserialize_model(self._model, serialized_parameters)
```
And I get that (the first row is for the original code and the second row is for the modified code) with the unbalanced data.
![s drawio](https://github.com/SMILELab-FL/FedLab/assets/70316218/3f3e9d70-97c8-4ef3-8b2d-cc97de005e53)

**Comment by AgentDS** [comment]
You can customize these ServerHandler. For vanilla FedAvg, you can use sample number to average the model weights. The code you showed above is what we used for our own experiment, actually.
